### PR TITLE
RBAC for Test Repositories - Issue #105

### PR DESCRIPTION
**Pull request by amizhen**
Resolves: https://github.com/AlmaLinux/build-system/issues/105
•Teams must now be assigned to Test Repositories upon creation
•Only members of a Test Repo's team can modify the Test Repo
•Done in conjunction with a pull request in frontend: https://github.com/amizhen/albs-frontend/tree/bs-105

### CODE REVIEW

File: alws/alembic.ini [hunk 1]
```diff
@@ -50,7 +50,7 @@ version_path_separator = os  # default: use os.pathsep
 # are written from script.py.mako
 # output_encoding = utf-8
 
-sqlalchemy.url = postgresql+psycopg2://postgres:password@db/almalinux-bs
+sqlalchemy.url = postgresql+psycopg2://postgres:password@db/db
```
Comment (javihernandez) [inline alws/alembic.ini:53]: I'd say you must keep this untouched

File: alws/alembic/versions/e952dc755f9d_add_team_to_test_repos.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+def upgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
+    op.create_table('test_repository_role_mapping',
+    sa.Column('test_repository_id', sa.Integer(), nullable=False),
+    sa.Column('role_id', sa.Integer(), nullable=False),
+    sa.ForeignKeyConstraint(['role_id'], ['user_roles.id'], name='fk_test_repositories_role_mapping_role_id', ondelete='CASCADE'),
+    sa.ForeignKeyConstraint(['test_repository_id'], ['test_repositories.id'], name='fk_test_repositories_role_mapping_product_id', ondelete='CASCADE'),
+    sa.PrimaryKeyConstraint('test_repository_id', 'role_id')
+    )
+    op.alter_column('build_releases', 'status',
```
Comment (javihernandez) [inline alws/alembic/versions/e952dc755f9d_add_team_to_test_repos.py:28]: We need to figure out why all these `nullable=False` are getting introduced in the migration. I bet is a side effect from recent migration to fastapi-sqla. @Korulag, any ideas?

File: alws/crud/test_repository.py [hunk 3]
```diff
@@ -139,9 +168,61 @@ async def bulk_create_package_mapping(
     await session.flush()
 
 
+def get_team_role_name(team_name: str, role_name: str):
+    return f'{team_name}_{role_name}'
+
+async def create_test_repository_role_mapping(session: AsyncSession, team_name: str):
+    # print('*MappingDB:0')
```
Comment (javihernandez) [inline alws/crud/test_repository.py:175]: get rid of print statements

File: alws/crud/test_repository.py [hunk 4]
```diff
@@ -139,9 +168,61 @@ async def bulk_create_package_mapping(
...
+    required_roles = (Contributor, Observer)
```
Comment (javihernandez) [inline alws/crud/test_repository.py:176]: I'd say we should maybe restrict to only team >= Contributor - Given that the Observers don't have the ability to perform any write action

File: alws/models.py [hunk 5]
```diff
@@ -823,6 +823,33 @@ def __repr__(self):
     ),
 )
 
+
+TestRepositoryRoleMapping = sqlalchemy.Table(
```
Comment (javihernandez) [inline alws/models.py:827]: Do we really need this?

File: alws/scripts/albs-gitea-listener/albs-gitea-listener-config.yaml [hunk 6]
```diff
@@ -1,12 +1,12 @@
 ---
```
Comment (javihernandez) [inline alws/scripts/albs-gitea-listener/albs-gitea-listener-config.yaml:1]: leave this file untouched

File: nginx_configs/albs.conf [hunk 7]
```diff
@@ -10,9 +10,9 @@ upstream pulp {
     server pulp:80;
```
Comment (javihernandez) [inline nginx_configs/albs.conf:10]: same, leave this file untouched

**Review by javihernandez — COMMENTED** [review]
Thanks @amizhen! Great job on your very first contribution to ALBS!
Overall, looks good, but needs some work. I left some review comments, none of them critical but must be addressed.

**Comment by javihernandez** [comment]
Regarding the tests failure of `tests.test_scripts.test_packages_exporter` ,  it is not related to your changes and it has been fixed in https://github.com/AlmaLinux/albs-web-server/commit/956257df51073db56ad9ec412c88714e85c3c576. All you need to do is to update your code with current master.

File: alws/crud/test_repository.py [hunk 8]
```diff
@@ -139,9 +168,61 @@ async def bulk_create_package_mapping(
...
+    #todo flush if flush see parameters
+    await session.flush()
+
+    return existing_roles
+
+
 async def create_repository(
     session: AsyncSession,
     payload: test_repository_schema.TestRepositoryCreate,
+    # team_id: int,
```
Comment (maccelf) [inline alws/crud/test_repository.py:225]: Remove this line

File: alws/crud/test_repository.py [hunk 9]
```diff
@@ -139,9 +168,61 @@ async def bulk_create_package_mapping(
...
 
+def get_team_role_name(team_name: str, role_name: str):
+    return f'{team_name}_{role_name}'
+
+async def create_test_repository_role_mapping(session: AsyncSession, team_name: str):
+    # print('*MappingDB:0')
+    required_roles = (Contributor, Observer)
+    required_actions = (UpdateTest, DeleteTest)
+    role_names = [get_team_role_name(team_name, role.name)
+                       for role in required_roles]
```
Comment (maccelf) [inline alws/crud/test_repository.py:179]: Wrong indent

File: alws/routers/test_repositories.py [hunk 10]
```diff
@@ -3,12 +3,14 @@
 from fastapi import APIRouter, Depends, HTTPException, status
 from fastapi_sqla import AsyncSessionDependency
 from sqlalchemy.ext.asyncio import AsyncSession
-
+# from alws.dependencies import get_db
```
Comment (maccelf) [inline alws/routers/test_repositories.py:6]: Delete this line

File: alws/crud/test_repository.py [hunk 11]
```diff
@@ -139,9 +168,61 @@ async def bulk_create_package_mapping(
...
+                actions_to_add.append(existing_action)
+
+    for role in existing_roles:
+        role_name = f'{team_name}_{role.name}'
+        # print('*MappingDB:', role_name)
+        for action in actions_to_add:
+            if action not in role.actions:
+                # print('*MappingDB:', role_name, action.name)
+                role.actions.append(action)
+            # else:
```
Comment (maccelf) [inline alws/crud/test_repository.py:212]: Remove all your unnecessary comments

File: alws/schemas/test_repository_schema.py [hunk 12]
```diff
@@ -45,6 +45,8 @@ class TestRepositoryCreate(BaseModel):
     url: AnyHttpUrlString
     tests_dir: str
     tests_prefix: typing.Optional[str] = None
+    team_id: int
+    owner_id: int
```
Comment (maccelf) [inline alws/schemas/test_repository_schema.py:49]: Why do we need `owner_id` when we create a test repository? I don't see where we use this variable

**Review by maccelf — CHANGES_REQUESTED** [review]

File: alws/schemas/test_repository_schema.py (hunk 12 above)
Comment (amizhen) [reply alws/schemas/test_repository_schema.py:49]: You are correct, I removed it and it worked fine

File: alws/crud/test_repository.py (hunk 11 above)
Comment (amizhen) [reply alws/crud/test_repository.py:212]: Done

File: alws/routers/test_repositories.py (hunk 10 above)
Comment (amizhen) [reply alws/routers/test_repositories.py:6]: Done

File: alws/crud/test_repository.py (hunk 9 above)
Comment (amizhen) [reply alws/crud/test_repository.py:179]: Done

File: alws/crud/test_repository.py (hunk 8 above)
Comment (amizhen) [reply alws/crud/test_repository.py:225]: Done

File: alws/models.py (hunk 5 above)
Comment (amizhen) [reply alws/models.py:827]: I think so. Its used here:
`class TestRepository(PermissionsMixin, TeamMixin, Base):
    __tablename__ = "test_repositories"
    id: Mapped[int] = mapped_column(sqlalchemy.Integer, primary_key=True)
    name: Mapped[str] = mapped_column(
        sqlalchemy.Text, nullable=False, unique=True
    )
    url: Mapped[str] = mapped_column(
        sqlalchemy.Text, nullable=False, unique=True
    )
    tests_dir: Mapped[str] = mapped_column(sqlalchemy.Text, nullable=False)
    tests_prefix: Mapped[Optional[str]] = mapped_column(
        sqlalchemy.Text, nullable=True
    )
    packages: Mapped[List["PackageTestRepository"]] = relationship(
        "PackageTestRepository",
        back_populates="test_repository",
        cascade="all, delete",
    )
    team: Mapped["Team"] = relationship("Team", back_populates="test_repositories")
    roles: Mapped[List["UserRole"]] = relationship(
        "UserRole", secondary=TestRepositoryRoleMapping
    )`

File: alws/crud/test_repository.py (hunk 3 above)
Comment (amizhen) [reply alws/crud/test_repository.py:175]: Done

File: alws/alembic/versions/e952dc755f9d_add_team_to_test_repos.py (hunk 2 above)
Comment (javihernandez) [reply alws/alembic/versions/e952dc755f9d_add_team_to_test_repos.py:28]: Ok, this seems to be a side-effect of the recent migration to sqlalchemy 2. I checked one by one in production db and all can be safely updated to set `nullable` to `False`, but some of them need to be handled:
* `build_tasks.mock_options`: these are old builds from 2021 - I'd say that we can update these with an empty `{}` or, if we don't want to, turn this into `nullable=True`
* `builds.mock_options`: these are old builds from 2021-2022 - I'd say that we can update these with a `{'definitions': {}}` or, if we don't want to, make it `nullable=True`
* `builds.cancel_testing ` - These are builds before the introduction of cancel_testing - can set those without cancel_testing with default value (False) 

@Korulag any thoughts? Also, would you like to make these adjustments within this PR? Or do you prefer to create another one?

File: alws/crud/test_repository.py [hunk 13]
```diff
@@ -139,6 +168,44 @@ async def bulk_create_package_mapping(
...
+    existing_roles = (
+        await session.execute(
+            select(UserRole).where(
+                UserRole.name.in_(role_names)
+            ).options(
+                selectinload(UserRole.actions)
+            )
+        )
+    ).scalars().all()
+    # existing_role_names = {r.name for r in existing_roles}
```
Comment (javihernandez) [inline alws/crud/test_repository.py:188]: Please delete this and other irrelevant comments

File: alws/schemas/test_repository_schema.py (hunk 12 above)
Comment (javihernandez) [reply alws/schemas/test_repository_schema.py:49]: @maccelf It's true that we shouldn't really need it, but look at `can_perform` implementation, which  [does the following](https://github.com/AlmaLinux/albs-web-server/blob/e842886770a6302053092d60e5b039b249d41835/alws/perms/authorization.py#L38C5-L38C31). If the object, in this case, the `test_repository` doesn't have an `owner_id`, `can_perform` will fail while trying to access `obj.owner.id`.  Or maybe I'm missing something here.

However, I'd say that we can stop sending it from frontend (is related to [this review comment](https://github.com/AlmaLinux/albs-frontend/pull/475#discussion_r1629717744) of the frontend side of this PR) and manually update the create payload with the current user that is actually performing the action.

File: alws/scripts/albs-gitea-listener/albs-gitea-listener-config.yaml (hunk 6 above)
Comment (javihernandez) [reply alws/scripts/albs-gitea-listener/albs-gitea-listener-config.yaml:1]: up

File: alws/crud/test_repository.py (hunk 13 above)
Comment (amizhen) [reply alws/crud/test_repository.py:188]: Done, I checked all of the other files, I believe I've removed them all.

File: alws/schemas/test_repository_schema.py (hunk 12 above)
Comment (javihernandez) [reply alws/schemas/test_repository_schema.py:49]: Ok, looking at this again, this is what we actually do when creating new products:
* frontend sends user_id -> https://github.com/AlmaLinux/albs-frontend/blob/master/src/pages/CreateProduct.vue#L127
* productCreate schema has owner_id -> https://github.com/AlmaLinux/albs-web-server/blob/master/alws/schemas/product_schema.py#L15
Said that, I'm not sure this is right, maybe we should use the team's owner for new created resources (either products or test_repositories). @maccelf, any thoughts on that?

Comment (soksanichenko) [reply alws/schemas/test_repository_schema.py:49]: > we should use the team's owner for new created resources (either products or test_repositories).

As I remember it's a core idea of RBAC model in ALBS when we was making a prototype.

Comment (maccelf) [reply alws/schemas/test_repository_schema.py:49]: IDK, Is it right if the non-owner of the team creates a test repository and then he will not be an owner of the test repository? Thats weird

Comment (javihernandez) [reply alws/schemas/test_repository_schema.py:49]: Well, that's quite a fair point. In my head, I'd say the resource should belong to the team, as the user is creating it on behalf of the team, but `owner_id` can't be a team, hence my proposal. But yes, let's follow the product's owner_id pattern (the user owns the resource). In that case, we do need this in the payload and this comment can be closed, right?

Comment (maccelf) [reply alws/schemas/test_repository_schema.py:49]: Yes, I think we can keep it as it is

**Pull request closed without merging**